pkg/network/ebpf/c/co-re/tracer-fentry.c

```diff
@@ -457,8 +457,24 @@ int BPF_PROG(tcp_connect, struct sock *sk) {
     return 0;
 }
 
+SEC("fexit/tcp_connect")
+int BPF_PROG(tcp_connect_exit, struct sock *sk, int rc) {
+    RETURN_IF_NOT_IN_SYSPROBE_TASK("fexit/tcp_connect");
+    if (rc == 0) {
+        // successful tcp_connect call, nothing to do
+        return 0;
+    }
+
+    // error returned from tcp_connect, remove entry
+    // from tcp_ongoing_connect_pid as tcp_finish_connect
+    // kprobe or tcp_close kprobe may not be called
+    // for this tcp_connect call
+    bpf_map_delete_elem(&tcp_ongoing_connect_pid, &sk);
+    return 0;
+}
+
 SEC("fentry/tcp_finish_connect")
-int BPF_PROG(tcp_finish_connect, struct sock *sk, struct sk_buff *skb, int rc) {
+int BPF_PROG(tcp_finish_connect, struct sock *sk, struct sk_buff *skb) {
```

Comment:
do we strictly need the tcp_finish_connect  if we're hooking the exit of tcp_connect ?

Reply:
`tcp_connect` just sends the `SYN` off, and gets the ball rolling for the connect. So the connect is not done when it returns. We're just checking now if the `tcp_connect` failed immediately, which is another case where `tcp_finish_connect` will not be called.

Reply:
Makes sense, thank you!